Purchases/Public/RCPurchases.m

```diff
@@ -534,32 +541,26 @@ - (void)        purchaseProduct:(SKProduct *)product
         return;
     }
 
-    if (!self.finishTransactions) {
```

Comment:
this is the trickiest file to understand the diff for, so I'd recommend doing side-by-side. 
Really, what I'm doing here is: 

- we had a bunch of one-liner `@synchronized` accesses. I don't think these were actually protecting us correctly - if two threads are both running this method, they're not guaranteed to stay consistent from one lock to the next. I moved to a single lock, where all of the state for the class is read / written. 
- I moved stuff that didn't need the lock and was just checking params to the top, before any method logic.

Reply:
this isn't necessary for this PR, but I'd spotted that before and figured we should fix it, given that there could be race conditions here.

Reply:
What are exactly the race conditions here? I am having some trouble understanding why running another thread on this method at the same time would break anything. The synchronized blocks were there because the accesses to `self.productsByIdentifier` and `self.presentedOfferingsByProductIdentifier` are not thread-safe so I understand those have to be synchronized, but why do we have to synchronize the accesses to `self.deviceCache` or `self.storeKitWrapper`

Reply:
re-reading this, I don't think there are currently race conditions. This is the last work day of the year, so my brain is a bit fried. Sorry for the noise. 

My thinking at first was that we have multiple properties of the same object being updated in a non-atomic way, so there might be places where we might be leaving the object in a temporarily inconsistent state. 

for example, we have the following objects: 
- `productsByIdentifier`
- `presentedOfferingsByProductIdentifier`
- `purchaseCompleteCallbacks`

each one is a mapping of stuff to product ids. 

So what I thought was that if the code makes any assumption like "if `purchaseCompleteCallbacks` has an entry for `productA`, then so must `productsByIdentifier`", this could break. 

However, we don't really have those assumptions coded in, since they're not really that related. So we should be safe without this change. 

The `storekitWrapper` bit really doesn't add a thing and could be removed. 
The `deviceCache` one, though, does add safety in that if between one access and the next, the value of `RCPurchases`'s `deviceCache` was changed (i.e.: the instance was replaced by a new one), you might have unexpected results. 
Say, if one `deviceCache` was initialized with a different `userDefaults` instance than the other, in between one access and the next, 💥 

In practice, that wouldn't happen with our current code, though.